// next-sanity/dist/index.js
import { createClient, unstable__adapter, unstable__environment } from "@sanity/client";
import { stegaClean } from "@sanity/client/stega";
import { createDataAttribute } from "@sanity/visual-editing/create-data-attribute";
import { jsx } from "react/jsx-runtime";
import { VisualEditing as VisualEditing$1 } from "./visual-editing.js";
import { defineLive } from "./_chunks-es/deprecated-live.js";
export * from "@portabletext/react";
import { defineQuery, default as default2 } from "groq";
import { isCorsOriginError } from "@sanity/next-loader";
let warned = !1;
function VisualEditing(props) {
  return warned || (console.warn(
    `Importing VisualEditing from the root import is deprecated and will be removed in next-sanity v11. Please change "import {VisualEditing} from 'next-sanity'" to "import {VisualEditing} from 'next-sanity/visual-editing'"`
  ), warned = !0), /* @__PURE__ */ jsx(VisualEditing$1, { ...props });
}
export {
  VisualEditing,
  createClient,
  createDataAttribute,
  defineLive,
  defineQuery,
  default2 as groq,
  isCorsOriginError,
  stegaClean,
  unstable__adapter,
  unstable__environment
};
