// next-sanity/dist/_chunks-es/NextStudioNoScript.js
import { jsx, jsxs } from "react/jsx-runtime";
import { memo } from "react";
const style = {
  height: "100vh",
  maxHeight: "100dvh",
  overscrollBehavior: "none",
  WebkitFontSmoothing: "antialiased",
  overflow: "auto"
}, NextStudioLayoutComponent = ({ children }) => /* @__PURE__ */ jsx("div", { id: "sanity", "data-ui": "NextStudioLayout", style, children }), NextStudioLayout = memo(NextStudioLayoutComponent), styles = {
  outer: {
    position: "absolute",
    top: 0,
    right: 0,
    left: 0,
    bottom: 0,
    background: "#fff",
    zIndex: 1
  },
  inner: {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    textAlign: "center",
    fontFamily: "helvetica, arial, sans-serif"
  }
}, NextStudioNoScript = () => /* @__PURE__ */ jsx("noscript", { children: /* @__PURE__ */ jsx("div", { style: styles.outer, children: /* @__PURE__ */ jsxs("div", { style: styles.inner, children: [
  /* @__PURE__ */ jsx("h1", { children: "JavaScript disabled" }),
  /* @__PURE__ */ jsxs("p", { children: [
    "Please ",
    /* @__PURE__ */ jsx("a", { href: "https://www.enable-javascript.com/", children: "enable JavaScript" }),
    " in your browser and reload the page to proceed."
  ] })
] }) }) });
export {
  NextStudioLayout,
  NextStudioNoScript
};
