// @portabletext/editor/lib/_chunks-cjs/util.is-selection-collapsed.cjs
"use strict";
var schema = require("@portabletext/schema"), util_sliceBlocks = require("./util.slice-blocks.cjs");
function getBlockEndPoint({
  context,
  block
}) {
  if (schema.isTextBlock(context, block.node)) {
    const lastChild = block.node.children[block.node.children.length - 1];
    if (lastChild)
      return {
        path: [...block.path, "children", {
          _key: lastChild._key
        }],
        offset: schema.isSpan(context, lastChild) ? lastChild.text.length : 0
      };
  }
  return {
    path: block.path,
    offset: 0
  };
}
function isEmptyTextBlock(context, block) {
  if (!schema.isTextBlock(context, block))
    return !1;
  const onlyText = block.children.every((child) => schema.isSpan(context, child)), blockText = util_sliceBlocks.getTextBlockText(block);
  return onlyText && blockText === "";
}
function isEqualSelectionPoints(a, b) {
  return a.offset === b.offset && JSON.stringify(a.path) === JSON.stringify(b.path);
}
function isSelectionCollapsed(selection) {
  return selection ? JSON.stringify(selection.anchor.path) === JSON.stringify(selection.focus.path) && selection.anchor.offset === selection.focus.offset : !1;
}
exports.getBlockEndPoint = getBlockEndPoint;
exports.isEmptyTextBlock = isEmptyTextBlock;
exports.isEqualSelectionPoints = isEqualSelectionPoints;
exports.isSelectionCollapsed = isSelectionCollapsed;
